allennlp/models/constituency_parser.py

```diff
@@ -249,21 +252,152 @@ def decode(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor
+                                                all_spans=all_spans,
+                                                num_spans=num_spans,
+                                                sentences=all_sentences,
+                                                pos_tags=all_pos_tags)
+
+        output_dict["trees"] = [trees[0] for trees in top_k_trees]
+        output_dict['top_k_trees'] = top_k_trees
         return output_dict
 
-    def construct_trees(self,
+    def compute_k_best(self,
+                       sentence: List[str],
+                       pos_tags: List[str],
+                       label_probabilities: torch.FloatTensor,
+                       span_to_index: Dict[Tuple[int, int], int],
+                       num_trees: List[int],
+                       distinguish_between_labels: bool = False):
+        """
+        :param sentence: The sentence for which top-k parses are being computed.
+        :param pos_tags: Part of speech tags for every token in the input sentence.
+        :param label_probabilities: A numpy array of shape (num_spans, num_labels)
+        :param span_to_index: A dictionary mapping span indices to column indices in
+        label_log_probabilities.
+        :param no_label_id: The id of the empty label.
+        :param num_trees: The number of parses required.
+        :param distinguish_between_labels: Whether to distinguish between different labels for the
+        top k trees.
+        :return: A list of the num_tree parses, and their log probabilities.
+        """
+        empty_label_index = self.vocab.get_token_index("NO-LABEL", "labels")
+        all_labels = [self.vocab.get_token_from_index(index, "labels").split("-") for index in
+                  range(self.vocab.get_vocab_size("labels"))]
+        label_probabilities_np = label_probabilities.cpu().numpy()
+        if not distinguish_between_labels:
+            temp = numpy.zeros((len(span_to_index), 2))
+            temp[:, 0] = label_probabilities_np[:, empty_label_index]
+            temp[:, 1] = 1 - label_probabilities_np[:, empty_label_index]
+
+            span_to_label = {}
+            # To ensure that the empty label does not have the maximum probability for any span.
+            label_probabilities_np[:, empty_label_index] = -1
+            for span, span_index in span_to_index.items():
+                label_index = label_probabilities_np[span_index, :].argmax()
+                span_to_label[span] = all_labels[label_index]
+
+            label_probabilities_np = temp
+            empty_label_index = 0
+
+
+        label_log_probabilities_np = numpy.log(label_probabilities_np + 10 ** -5)
+        correction_term = numpy.sum(label_log_probabilities_np[:, empty_label_index])
+        label_log_probabilities_np -= label_log_probabilities_np[:, empty_label_index]\
+            .reshape((len(span_to_index), 1))
+        cache = {}
+
+        def helper(left, right, must_be_constituent):
```

Comment:
this name is unhelpful - can you refactor this helper function to contain a lot more comments, like the previous tree assembly code? This is the most complicated function in this entire model and right now it's pretty unreadable.
https://github.com/allenai/allennlp/pull/1100/files?diff=split#diff-d2b9b626f33d5fb4c5fa76a087826bb9L405